scaletest/trafficgen/run.go

```diff
@@ -0,0 +1,281 @@
+		conn.Close()
+		close(rch)
+	}()
+
+	// Write random data to the PTY every tick.
+	go func() {
+		logger.Debug(ctx, "writing to agent", slog.F("agent_id", agentID))
+		wch <- writeRandomData(deadlineCtx, &crw, bytesPerTick, tick.C)
+		logger.Debug(ctx, "done writing to agent", slog.F("agent_id", agentID))
+		close(wch)
+	}()
+
+	// Write until the context is canceled.
+	if wErr := <-wch; wErr != nil {
+		return xerrors.Errorf("write to pty: %w", wErr)
+	}
+	if rErr := <-rch; rErr != nil {
+		return xerrors.Errorf("read from pty: %w", rErr)
+	}
+
+	duration := time.Since(start)
+
+	logger.Info(ctx, "results",
+		slog.F("duration", duration),
+		slog.F("sent", crw.BytesWritten()),
+		slog.F("rcvd", crw.BytesRead()),
+	)
+
+	return nil
+}
+
+// Cleanup does nothing, successfully.
+func (*Runner) Cleanup(context.Context, string) error {
+	return nil
+}
+
+// drainContext drains from src until it returns io.EOF or ctx times out.
+func drainContext(ctx context.Context, src io.Reader) error {
+	errCh := make(chan error, 1)
+	done := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				_, err := io.CopyN(io.Discard, src, 1)
+				if ctx.Err() != nil {
+					return // context canceled while we were copying.
+				}
+				if err != nil {
+					errCh <- err
+					close(errCh)
+					return
+				}
+			}
+		}
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			close(done)
+			return nil
+		case err := <-errCh:
+			if err != nil {
+				if xerrors.Is(err, io.EOF) {
+					return nil
+				}
+				// It's OK if the context is canceled.
+				if xerrors.Is(err, context.DeadlineExceeded) {
+					return nil
+				}
+				return err
+			}
+		}
+	}
+}
+
+func writeRandomData(ctx context.Context, dst io.Writer, size int64, tick <-chan time.Time) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-tick:
+			payload := "#" + mustRandStr(size-1)
+			data, err := json.Marshal(codersdk.ReconnectingPTYRequest{
+				Data: payload,
+			})
+			if err != nil {
+				return err
+			}
+			if _, err := copyContext(ctx, dst, data); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// copyContext copies from src to dst until ctx is canceled.
+func copyContext(ctx context.Context, dst io.Writer, src []byte) (int, error) {
+	var count int
+	for {
+		select {
+		case <-ctx.Done():
+			return count, nil
+		default:
+			for idx := range src {
+				n, err := dst.Write(src[idx : idx+1])
+				if err != nil {
+					if xerrors.Is(err, io.EOF) {
+						return count, nil
+					}
+					if xerrors.Is(err, context.DeadlineExceeded) {
+						// It's OK if we reach the deadline before writing the full payload.
+						return count, nil
+					}
+					return count, err
+				}
+				count += n
+			}
+			return count, nil
+		}
+	}
+}
+
+// countReadWriter wraps an io.ReadWriter and counts the number of bytes read and written.
+type countReadWriter struct {
+	ctx context.Context
```

Comment:
ctx seems to be used to intercept IO on context expiration.  This is unnecessary since we close the underlying `net.Conn` when the context expires.

Reply:
Removed!